java/arcs/core/data/TypeUtils.kt

```diff
@@ -0,0 +1,140 @@
+      SchemaRegistry.getSchema(fieldType.schemaHash)
+    )
+    is FieldType.Tuple -> mapTupleToInferredType(fieldName, fieldType)
+    is FieldType.ListOf -> mapListTypeToInferredType(fieldType)
+    is FieldType.InlineEntity -> mapSchemaToInferredType(
+      SchemaRegistry.getSchema(fieldType.schemaHash)
+    )
+  }
+}
+
+private fun mapListTypeToInferredType(fieldType: FieldType.ListOf) =
+  InferredType.SeqType(
+    when (fieldType.primitiveType) {
+      is FieldType.InlineEntity -> mapSchemaToInferredType(
+        SchemaRegistry.getSchema((fieldType.primitiveType as FieldType.InlineEntity).schemaHash)
+      )
+      else -> mapPrimitiveTypeToInferredType(fieldType.primitiveType)
+    }
+  )
+
+private fun mapSchemaToInferredType(schema: Schema): InferredType {
+  val allFields = schema.fields.singletons.entries + schema.fields.collections.entries
+  return InferredType.ScopeType(
+    MapScope(
+      schema.names.firstOrNull()?.name ?: "*",
+      allFields.map { (field, ftype) ->
+        val inferredType = mapFieldTypeToInferredType(field, ftype)
+        field to if (field in schema.fields.singletons) inferredType else InferredType.SeqType(
+          inferredType
+        )
+      }.associateBy({ it.first }, { it.second })
+    )
+  )
+}
+
+/** VisibleForTesting */
+fun mapTypeToInferredType(type: Type): InferredType = when (type) {
+  is CollectionType<*> -> InferredType.SeqType(mapTypeToInferredType(type.containedType))
+  is Type.TypeContainer<*> -> mapTypeToInferredType(type.containedType)
+  is EntityType -> mapSchemaToInferredType(type.entitySchema)
+  else -> throw IllegalArgumentException("Can't map type $type to InferredType")
+}
+
+/** Map [Type] to [InferredType] recursively creating sub-scopes as needed. */
+fun constructTypeScope(map: Map<String, Type>): MapScope<InferredType> {
+  return MapScope("root", map.mapValues { (_, type) ->
+    mapTypeToInferredType(
+      type
+    )
+  })
+}
+
+/**
+ * Typecheck a paxel [Expression] against a [Scope] of [InferredType].
+ *
+ * @param [name] name of the connection being checked.
+ * @param [expr] is the expression to check.
+ * @param [typeScope] usually from [constructTypeScope] from readable HandleConnection connections.
+ * @param [param] is a Scope of [InferredType] for any query parameters the expression may refer to.
+ * @param [expectedType] [HandleConnection] [Type] the Paxel expression should be assignable to.
+ * @return a list of warnings if succeeded
+ * @throws PaxelTypeException if any errors occur.
+ */
+fun typeCheck(
+  name: String,
+  expr: Expression<*>,
+  typeScope: Expression.Scope,
+  expectedType: Type,
+  params: Expression.Scope = ParameterScope()
+): List<String> {
+  val typeEvaluator = TypeEvaluator(params)
+  try {
+    val inferredType = expr.accept(typeEvaluator, typeScope)
+    val expectedInferredType = mapTypeToInferredType(expectedType)
+    if (!expectedInferredType.isAssignableFrom(inferredType)) {
```

Comment:
This will fail with a hard error. You can't assign {name} with {name,age}. The LHS is ScopeType(Scope("name" to TextType)) the RHS is ScopeType(Scope("name" to TextType, "age" to "IntType")

Maybe you mean we should warn in this case
```
particle Foo
  foo: writes Person {name: Text, age: Int} = new Person {
    name: "John",
  }
```

In this case, isAssignableFrom() returns true, but you didn't set the 'Age' field, so either you want the default, or your handle connection is overly broad declaration.

Reply:
so I feel the case you provided should be an error, or at least a warning.

I just checked `isAssignableFrom` on a ScopeType and it relies on `scope == other.scope`, which would fall back to map equality, so it would fail I think if LHS and RHS scopes have different entries, no?

Reply:
Good catch, you're right, ScopeType.isAssignableFrom() is wrong, as {a,b} = {a} should work, but {a} = {a,b} should fail. Right now, only {a,b} = {a,b} works. I will fix and add coverage.